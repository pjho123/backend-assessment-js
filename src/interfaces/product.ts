import { IRequestStrict } from "itty-router"

export interface IProduct {
  id: number,
  title: string,
  tags: string,
  created_at: string,
  updated_at: string,
  sku: string,
}

interface IThirdPartyProductVariant {
  id: number,
  title: string,
  sku: string,
  created_at: string,
  updated_at: string,
}

export interface IThirdPartyProduct {
  title: string,
  tags: string,
  variants: IThirdPartyProductVariant[]
}

export interface IThirdPartyResponse {
  products: IThirdPartyProduct[],
}

export type TDeleteProductRequest = {
  product_id: number
} & Request;
