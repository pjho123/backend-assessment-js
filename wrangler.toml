name = "ma-project"
main = "src/index.ts"
compatibility_date = "2023-04-18"
workers_dev = false
compatibility_flags = ["experimental:nodejs_compat_v2"]

[vars]
PGHOST = "ep-cold-meadow-a78on0ae-pooler.ap-southeast-2.aws.neon.tech"
PGDATABASE = "assessment-db"
PGUSER = "assessment-db_owner"
PGPASSWORD = "CnkBFR9sa2df"
ENDPOINT_ID = "ep-cold-meadow-a78on0ae-pooler"
API_KEY = "c6c687a0a7209f877f89a40824907ae5"