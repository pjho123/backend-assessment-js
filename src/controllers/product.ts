import { error } from "itty-router";
import { TDeleteProductRequest } from "../interfaces/product";
import ProductFormatter from "../lib/product-formatter";
import ThirdPartyDataSource from "../lib/thirdparty-datasource";
import ProductService from "../services/product-service";
import dayjs from "dayjs";

class ProductsController {
  static async get(): Promise<Response> {
    const data = await ThirdPartyDataSource.getData();
    const transformedProducts = ProductFormatter.format(data);

    await ProductService.createBulk(transformedProducts);

    return Response.json({ message: 'Third Party Data has been added to the DB.' });
  }

  static async create(): Promise<Response> {
    const data = await ThirdPartyDataSource.getProductData();
    const transformedProducts = ProductFormatter.format(data);

    await ProductService.createBulk(transformedProducts);

    const products = await ProductService.getByIds(transformedProducts.map(d => d.id));
    return Response.json({
      products: products.map(p => ({
        ProductID: p.id,
        Title: p.title,
        Tags: p.tags,
        CreatedAt: dayjs(p.created_at).toString(),
        UpdatedAt: dayjs(p.updated_at).toString(),
        ProductCode: p.sku,
      }))
    });
  }

  static async update(): Promise<Response> {
    const products = await ProductService.getAll();

    const result = [];
    for (const product of products) {
      let updatedProduct = product;

      if (!product.title.includes(product.sku)) {
        updatedProduct = await ProductService.updateProductById(product.id, {
          title: `${product.title} ${product.sku}`,
        });
      }

      result.push(updatedProduct);
    }

    return Response.json({
      products: result.map(p => ({
        ProductID: p.id,
        Title: p.title,
        Tags: p.tags,
        CreatedAt: dayjs(p.created_at).toString(),
        UpdatedAt: dayjs(p.updated_at).toString(),
        ProductCode: p.sku,
      }))
    });
  }

  static async delete({ product_id }: TDeleteProductRequest): Promise<Response> {
    const product = await ProductService.getById(product_id);

    if (!product) {
      return error(400, 'Product not found.');
    }

    await ProductService.delete(product_id);

    return Response.json({ message: 'Product has been successfully deleted.' });
  }
}

export default ProductsController;
