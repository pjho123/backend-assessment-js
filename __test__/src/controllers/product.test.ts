import { unstable_dev } from "wrangler";
import type { UnstableDevWorker } from "wrangler";
import { describe, expect, it, beforeAll, afterAll, vi } from "vitest";
import ProductService from "../../../src/services/product-service";
import ProductsController from "../../../src/controllers/product";
import ThirdPartyDataSource from "../../../src/lib/thirdparty-datasource";
import { IProduct, TDeleteProductRequest } from "../../../src/interfaces/product";

describe("Worker", () => {
	let worker: UnstableDevWorker;

	beforeAll(async () => {
		worker = await unstable_dev("src/index.ts", {
			experimental: { disableExperimentalWarning: true },
		});
		vi.mock("./person", () => {
			const Person = vi.fn(() => ({
				getAge: vi.fn(() => 15),
			}));
			return { Person };
		});
		// vi.mock('../../../src/services/product-service', () => {
		// 	// const ProductService = vi.fn();
		// 	// ProductService.prototype.createBulk = vi.fn(() => 15);
		// 	const ProductService = vi.fn(() => ({
		// 		getAll: vi.fn().mockResolvedValue(true),
		// 		getById: vi.fn().mockResolvedValue(true),
		// 		getByIds: vi.fn().mockResolvedValue(true),
		// 		updateProductById: vi.fn().mockResolvedValue(true),
		// 		createBulk: vi.fn().mockResolvedValue(true),
		// 		delete: vi.fn().mockResolvedValue(true),
		// 	}));
		// 	return { default: ProductService };
		// });
		const product: IProduct = {
			id: 11111,
			title: 'Sample Product',
			tags: 'tags',
			created_at: '',
			updated_at: '',
			sku: 'sample-sku',
		};
		vi.spyOn(ProductService, 'getAll').mockResolvedValue([product]);
		vi.spyOn(ProductService, 'getById').mockResolvedValue(product);
		vi.spyOn(ProductService, 'getByIds').mockResolvedValue([product]);
		vi.spyOn(ProductService, 'updateProductById').mockResolvedValue(product);
		vi.spyOn(ProductService, 'createBulk').mockResolvedValue([product]);
		vi.spyOn(ProductService, 'delete').mockResolvedValue(true);

		vi.spyOn(ThirdPartyDataSource, 'getData').mockResolvedValue([{
			title: 'Sample Product 1',
			tags: 'tag1',
			variants: [{
				id: 111111,
				title: 'variant-title1',
				sku: 'SKU1',
				updated_at: '2024-07-28T20:15:18-04:00',
				created_at: '2024-07-28T20:14:32-04:00',
			}]
		}]);
		vi.spyOn(ThirdPartyDataSource, 'getProductData').mockResolvedValue([{
			title: 'Sample Product 2',
			tags: 'tag2',
			variants: [{
				id: 22222,
				title: 'variant-title2',
				sku: 'SKU2',
				updated_at: '2024-07-28T20:15:18-04:00',
				created_at: '2024-07-28T20:14:32-04:00',
			}]
		}]);
	});

	afterAll(async () => {
		await worker.stop();
	});

	describe('Controller for GET /api/products', () => {
		it("should return 200", async () => {
			const resp = await ProductsController.get();

			const jsonResponse = await resp.json();

			expect(resp.status).toBe(200);
			expect(jsonResponse).toMatchSnapshot();
		});
	});

	describe('Controller for POST /api/products', () => {
		it("should return 200", async () => {
			const resp = await ProductsController.create();

			const jsonResponse = await resp.json();

			expect(resp.status).toBe(200);
			expect(jsonResponse).toMatchSnapshot();
		});
	});

	describe('Controller for PUT /api/products', () => {
		it("should return 200", async () => {
			const resp = await ProductsController.update();

			const jsonResponse = await resp.json();

			expect(resp.status).toBe(200);
			expect(jsonResponse).toMatchSnapshot();
		});
	});

	describe('Controller for DELETE /api/products', () => {
		it("should return 200", async () => {
			// @ts-ignore
			const resp = await ProductsController.delete({ product_id: 12345 });

			const jsonResponse = await resp.json();

			expect(resp.status).toBe(200);
			expect(jsonResponse).toMatchSnapshot();
		});
	});
});
