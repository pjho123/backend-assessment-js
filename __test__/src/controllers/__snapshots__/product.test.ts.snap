// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Worker > Controller for DELETE /api/products > should return 200 1`] = `
{
  "message": "Product has been successfully deleted.",
}
`;

exports[`Worker > Controller for GET /api/products > should return 200 1`] = `
{
  "message": "Third Party Data has been added to the DB.",
}
`;

exports[`Worker > Controller for POST /api/products > should return 200 1`] = `
{
  "products": [
    {
      "CreatedAt": "Sat, 21 Sep 2024 10:08:26 GMT",
      "ProductCode": "sample-sku",
      "ProductID": 11111,
      "Tags": "tags",
      "Title": "Sample Product",
      "UpdatedAt": "Sat, 21 Sep 2024 10:08:26 GMT",
    },
  ],
}
`;

exports[`Worker > Controller for PUT /api/products > should return 200 1`] = `
{
  "products": [
    {
      "CreatedAt": "Sat, 21 Sep 2024 10:08:26 GMT",
      "ProductCode": "sample-sku",
      "ProductID": 11111,
      "Tags": "tags",
      "Title": "Sample Product",
      "UpdatedAt": "Sat, 21 Sep 2024 10:08:26 GMT",
    },
  ],
}
`;
